### PR TITLE
[SPARK-46739][SQL] Add the error class `UNSUPPORTED_CALL`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to add new error class for unsupported method calls, and remove similar legacy error classes. New `apply()` method of `SparkUnsupportedOperationException` extracts method and class name from stack traces automatically, and places them to error class parameters.

### Why are the changes needed?
To improve code maintenance by avoid boilerplate code (extract class and method names automatically), and to clean up `error-classes.json`. 

### Does this PR introduce _any_ user-facing change?
Yes, it can if user's code depends on the error class or message format of `SparkUnsupportedOperationException`.

### How was this patch tested?
By running new test:
```
$ build/sbt "test:testOnly *QueryCompilationErrorsSuite"
```
and the affected test suites:
```
$ build/sbt "core/testOnly *SparkThrowableSuite"
$ build/sbt "test:testOnly *ShuffleSpecSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: common/utils/src/main/scala/org/apache/spark/SparkException.scala [hunk 1]
```diff
@@ -228,6 +228,19 @@ private[spark] class SparkUnsupportedOperationException private(
   override def getErrorClass: String = errorClass.orNull
 }
 
+private[spark] object SparkUnsupportedOperationException {
+  def apply(): SparkUnsupportedOperationException = {
+    val stackTrace = Thread.currentThread().getStackTrace
+    val messageParameters = if (stackTrace.length >= 4) {
```
Comment (MaxGekk) [inline common/utils/src/main/scala/org/apache/spark/SparkException.scala:234]: The stack trace looks like:
Screenshot 2024-01-16 at 19 55 35

**Comment by MaxGekk** [comment]
@cloud-fan @beliefer @LuciferYang @panbingkun Could you take a look at the PR, please.

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @cloud-fan for review.

**Comment by LuciferYang** [comment]
late LGTM

**Pull request closed without merging**